src/common/column_matrix.h

```diff
@@ -103,96 +110,179 @@ class ColumnMatrix {
-        accum_row_ind_ += static_cast<size_t>(nrow);
       } else {
         accum_index_ += feature_counts_[fid];
         accum_row_ind_ += feature_counts_[fid];
       }
       boundary_[fid].index_end = accum_index_;
-      boundary_[fid].row_ind_end = accum_row_ind_;
     }
 
-    index_.resize(boundary_[nfeature - 1].index_end);
-    row_ind_.resize(boundary_[nfeature - 1].row_ind_end);
+    if ( (gmat.max_num_bins_ - 1) <= static_cast<int>(std::numeric_limits<uint8_t>::max()) ) {
+      type_size_ = 1;
+    } else if ((gmat.max_num_bins_ - 1) <= static_cast<int>(std::numeric_limits<uint16_t>::max())) {
+      type_size_ = 2;
+    } else {
+      type_size_ = 4;
+    }
 
-    // store least bin id for each feature
-    index_base_.resize(nfeature);
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      index_base_[fid] = gmat.cut.Ptrs()[fid];
+    index_.resize(boundary_[nfeature - 1].index_end * type_size_, 0);
+    if (!all_dense) {
+      row_ind_.resize(boundary_[nfeature - 1].index_end);
     }
 
+    // store least bin id for each feature
+    index_base_ = const_cast<uint32_t*>(gmat.cut.Ptrs().data());
+
     // pre-fill index_ for dense columns
+    const bool noMissingValues = gmat.row_ptr[nrow] == nrow * nfeature;
 
-    #pragma omp parallel for
-    for (int32_t fid = 0; fid < nfeature; ++fid) {
-      if (type_[fid] == kDenseColumn) {
-        const size_t ibegin = boundary_[fid].index_begin;
-        uint32_t* begin = &index_[ibegin];
-        uint32_t* end = begin + nrow;
-        std::fill(begin, end, std::numeric_limits<uint32_t>::max());
-        // max() indicates missing values
+    if (noMissingValues) {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, false);
+    } else {
+      missing_flags_.resize(boundary_[nfeature - 1].index_end, true);
+    }
+
+    if (all_dense) {
+      switch (gmat.index.getBinBound()) {
+        case POWER_OF_TWO_8:
+          SetIndexAllDense(gmat.index.data<uint8_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_16:
+          SetIndexAllDense(gmat.index.data<uint16_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+        case POWER_OF_TWO_32:
+          SetIndexAllDense(gmat.index.data<uint32_t>(), gmat, nrow, nfeature, noMissingValues);
+          break;
+      }
+    } else {
+      switch (type_size_) {
+        case 1:
+          SetIndex<uint8_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 2:
+          SetIndex<uint16_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+        case 4:
+          SetIndex<uint32_t>(gmat.index.data<uint32_t>(), gmat, nrow, nfeature);
+          break;
+      }
+    }
+  }
+
+  /* Fetch an individual column. This code should be used with XGBOOST_TYPE_SWITCH
+     to determine type of bin id's */
+  template <typename T>
+  inline Column<T> GetColumn(unsigned fid) const {
+    Column<T> c(type_[fid],
```

Comment:
Is it possible to clean up the `Column` a little bit?  I find it very difficult to track.  Please try to dumb it down for others. ;-)

Reply:
yes, I`ll try to clean up full `Column` class